server/artifacts/artifact_server.go

```diff
@@ -81,6 +83,174 @@ func (a *ArtifactServer) GetInputArtifact(w http.ResponseWriter, r *http.Request
 	a.getArtifact(w, r, true)
 }
 
+// single endpoint to be able to handle serving directories as well as files, both those that have been archived and those that haven't
+// Valid requests:
+//  /artifact-files/{namespace}/[archived-workflows|workflows]/{id}/{nodeId}/outputs/{artifactName}
+//  /artifact-files/{namespace}/[archived-workflows|workflows]/{id}/{nodeId}/outputs/{artifactName}/{fileName}
+//  /artifact-files/{namespace}/[archived-workflows|workflows]/{id}/{nodeId}/outputs/{artifactName}/{fileDir}/.../{fileName}
+// 'id' field represents 'uid' for archived workflows and 'name' for non-archived
+func (a *ArtifactServer) GetArtifactFile(w http.ResponseWriter, r *http.Request) {
+
+	const (
+		NAMESPACE_INDEX       = 2
+		ARCHIVE_DISCRIM_INDEX = 3
+		ID_INDEX              = 4
+		NODE_ID_INDEX         = 5
+		DIRECTION_INDEX       = 6
+		ARTIFACT_NAME_INDEX   = 7
+		FILE_NAME_FIRST_INDEX = 8
+	)
+
+	var fileName *string
+	requestPath := strings.Split(r.URL.Path, "/")
```

Comment:
minor you could use `SplitN`, this will put all trailing part, saving the join

Reply:
Not sure if I understand what you mean by the advantage, but the reason I didn't use SplitN was because the number is undefined: could be "../artifactName", "../artifactName/filename", "../artifactName/filedir/filename", "../artifactName/filedir1/filedir2/filename", etc

Reply:
if you SplitN() then you can:

```golang
fileName = requestPath[fileNameFirstIndex]
```

no need to join

Reply:
I see. And it looks like if you request n, it's "at most" n, so okay if you only have n - 1. Good to know.